### DISCUSSION TITLE
We should have an `onCancelled` and `onLogin` event

### DISCUSSION THREAD

**Discussion by ba32107** [Glench/ExtPay#234] [Ideas]
Hi! 

CurrentIy, the `onPaid` event is fired both when a user signs up, and when a user restores an existing subscription. The docs say:

> Note: onPaid callbacks will be called after a user pays as well as after a user "logs in" (e.g. activates their paid account on a different browser/profile/install). This may change in the future -- if you'd like this to work differently, please contact me with a detailed explanation of your use case :)

I believe there should be two different events for this. My use-case: I want to send analytics data on both these events, and currently the events triggered by logins skew the signed up count. So for example, if a user pays once, then reinstalls the extension 10 times and logs in each time, I will get 11 "paid" analytics events. I know it's an extreme example, but still important. (actually, I don't really care about the login event, but I do care about the "paid" event only firing when the user actually pays)

Related, there is no way to detect when the user cancels the subscription. I know the plan isn't canceled straight away, but we should know when the user makes the decision to not pay for our product anymore. I spent some time today reading the ExtPay code and understanding how these event handlers work, and I believe it would be pretty easy to add this event. Just need to fire a message when the user clicks the "cancel" button in the ExtPay popup. Currently my only alternative is to somehow poll the user in my background script and check the `subscriptionCancelAt` field.

Tracking these events correctly is very important, since all of us developers here want to monetize our products. I am building analytics around user conversion rate and cancellations, and the second one is a blind spot.

PS: I would be happy to contribute! I have a reasonable understanding of how the open source part works. However, in order to contribute these events, we also need to change your server side, so I am not able to help :)

**Comment by ba32107** [comment]
In the interim, I would also be happy with a reasonable workaround for the cancelled event. My only idea so far is, whenever I open the payments page for the user, I will poll the user's `subscriptionCancelAt` field for, say, 5 minutes? The idea being that if someone wants to cancel, they have to open the page, and hopefully they will complete it within 5 minutes.

**Reply by Glench** [reply]
Please DO NOT poll the `subscriptionCancelAt` field. This is a great way to DDOS ExtensionPay's servers. What you can do instead is whenever you would normally call `extpay.getUser()` within your extension, replace it with your own function that calls `extpay.getUser()` and also checks if the `subscriptionCancelAt` field changes so it's a more on-demand check. OR you can add your own DOM listener to the cancel button in the ExtensionPay popup.

Otherwise, the `onLogin` and `onCancel` events are a good idea. Right now we're focused on getting ExtPay 4.0 done, but maybe we'll add those after. Cheers!

**Reply by ba32107** [reply]
> OR you can add your own DOM listener to the cancel button in the ExtensionPay popup.

This is a good idea and seems the most straightforward - will do this, thanks! Don't know why I didn't think of this one :) The only issue is that my implementation will depend on the DOM structure of your popup, so on the off chance you change that, my code will break. But it's a non-critical path, so it's fine.

**Reply by ba32107** [reply]
Hi @Glench , do you have a suggestion for a workaround to implement `onLogin`? I was able to detect the login event by putting this into a content script:
```Javascript
window.addEventListener('message', (event) => {
    if (event.origin !== 'https://extensionpay.com' || event.source !== window || event.data !== "fetch-user") {
        return
    }

    if (event.source.location.href.toLowerCase().includes('reactivate-from-email')) {
        console.log('login!')
    }

}, false)
```

This works fine, I can track the login event here (for my analytics). However this event will still reach the `onPaid` event listener (which is where I'm tracking the "user signed up" analytics) so it will still count logins as "subscribed" events which is not what I want. Do you have a good idea here?

The only thing I can think of is to completely ditch the `onPaid` event listener, and extend the above code with something like this:
```Javascript
    if (event.source.location.href.toLowerCase().includes('reactivate-from-email')) {
        console.log('track login event')
    } else if (event.source.location.href.toLowerCase().includes('subscribed')) {
        console.log('track new subscriber event')
    }
```

But I don't really like this, because now I'm completely dependent on your internal implementation for my analytics. I would like to use the `onPaid` event which is part of your public stable API, if I can.
